crates/pathfinder/src/state/sync.rs

```diff
@@ -138,24 +121,35 @@ where
 
     loop {
         tokio::select! {
+            l1_event = rx_l1.recv() => match l1_event {
+                Some(update) => {
+                    let tx = db_conn.transaction().context("db tx")?;
+
+                    let l1_l2_head = RefsTable::get_l1_l2_head(&tx)?.map(|x| x.0).unwrap_or_default();
+                    if update.block_number <= l1_l2_head {
+                        tracing::info!(block=update.block_number, "L1 reorg detected");
+                    }
+                    RefsTable::set_l1_l2_head(&tx, Some(StarknetBlockNumber(update.block_number)))?;
+                    tracing::info!(block=update.block_number, "L1-L2 head update");
+
+                    if let Some(block) = StarknetBlocksTable::get(&tx, StarknetBlocksBlockId::Number(StarknetBlockNumber(update.block_number)))? {
+                        if block.root.0.as_ref() != update.global_root.as_bytes() {
+                            tracing::warn!(block=update.block_number, "Block root mismatch (L1={}, DB={})", update.global_root, block.root);
+                            // TODO(SM): shut down? ignore?
+                        }
+                    }
+
+                    tx.commit().context("db tx commit")?;
+                },
+                None => {
+                    tracing::warn!("L1 sync stop");
+                    return Ok(());
```

Comment:
I think this means the node will stop if L1 sync task fails? Probably not what we want. L1 sync task should be restarted similar to how `main` currently does this (with a small restart delay).

Reply:
I made `l1_sync` "infallible" (returning `()` instead of `anyhow::Result<()>`), thus no need to restart the L1 sync. So now it is effectively unreachable, as channel being dropped from the main loop means that whole runtime is being dropped.

Basically a trade-off between "restarting boilerplate and using `?`" vs. "loop never exits". I want to suggest that for L2 as well (remove "restart"), but not in this PR.

Reply:
Changed to `unreachable!()`

Reply:
Maybe I'm misunderstanding but does that not mean if the L1 loop ever panics or fails that the node crashes?